### PR TITLE
Add AuthHashService; refactor code to use it.

### PR DESCRIPTION
**Pull request by demiankatz**
This also adds test coverage for the EmailAuthenticator and fixes some existing minor issues (such as use of some inappropriate types) that are not directly related to the refactoring.

TODO
- [x] Port ExpireAuthHashesCommand

### CODE REVIEW

**Comment by demiankatz** [comment]
I believe this is now ready for review. A couple of notes:

1.) I don't have an easy way to actually test this functionality in my test environment, so all of the work here is based on the test suite (and I significantly expanded test coverage as part of what I've done). If you can test that this _actually_ works, that would be a very big help!

2.) I've revised the abstract expiration command here to work with either tables or services, and I've defined a feature interface for services so we can simplify some of the validation logic in the future. I don't think it will take much work at all to migrate the remaining expiration commands to use this new mechanism. That would be a great follow-up action once this is merged, but I'll wait and see what you think of the refactoring before I extend it further.

File: module/VuFind/src/VuFind/Db/Service/AuthHashService.php [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    /**
+     * Delete an auth_hash entity object.
+     *
+     * @param AuthHashEntityInterface|int $authHashOrId Object or ID value representing auth_hash to delete
+     *
+     * @return void
+     */
+    public function deleteAuthHash(AuthHashEntityInterface|int $authHashOrId)
```
Comment (aleksip) [inline module/VuFind/src/VuFind/Db/Service/AuthHashService.php]: `: void`

File: module/VuFind/src/VuFind/Db/Service/AuthHashServiceInterface.php [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    public function createEntity(): AuthHashEntityInterface;
+
+    /**
+     * Delete an auth_hash entity object.
+     *
+     * @param AuthHashEntityInterface|int $authHashOrId Object or ID value representing auth_hash to delete
+     *
+     * @return void
+     */
+    public function deleteAuthHash(AuthHashEntityInterface|int $authHashOrId);
```
Comment (aleksip) [inline module/VuFind/src/VuFind/Db/Service/AuthHashServiceInterface.php]: `: void`

File: module/VuFind/tests/unit-tests/src/VuFindTest/Auth/EmailAuthenticatorTest.php [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+    /**
+     * Get an EmailAuthenticator to test.
+     *
+     * @param ?SessionManager           $sessionManager  Session manager
+     * @param ?CsrfInterface            $csrf            CSRF validator
+     * @param ?Mailer                   $mailer          Mailer service
+     * @param ?PhpRenderer              $renderer        View renderer
+     * @param ?RemoteAddress            $remoteAddress   Remote address details
+     * @param array                     $config          Configuration settings
+     * @param ?AuthHashServiceInterface $authHashService AuthHash daabase service
```
Comment (aleksip) [inline module/VuFind/tests/unit-tests/src/VuFindTest/Auth/EmailAuthenticatorTest.php]: database

File: module/VuFindConsole/src/VuFindConsole/Command/Util/AbstractExpireCommand.php [hunk 4]
```diff
@@ -81,24 +83,26 @@ class AbstractExpireCommand extends Command
...
+    public function __construct(
+        protected Gateway|DeleteExpiredInterface $service,
+        ?string $name = null
+    ) {
+        if (!method_exists($service, 'deleteExpired')) {
+            $serviceName = $service::class;
+            throw new \Exception("$serviceName does not support deleteExpired()");
         }
-        $this->table = $table;
+        $this->service = $service;
```
Comment (aleksip) [inline module/VuFindConsole/src/VuFindConsole/Command/Util/AbstractExpireCommand.php]: Not needed for promoted property.

**Review by aleksip — APPROVED** [review]

**Pull request merged**